uitools/toolkitcpp/cpp/Esri/ArcGISRuntime/Toolkit/ArcGISAuthenticationController.cpp

```diff
@@ -0,0 +1,263 @@
+#include <QPointer>
+#include <QUrl>
+
+// Maps SDK headers
+#include <ArcGISAuthenticationChallenge.h>
+#include <ArcGISAuthenticationManager.h>
+#include <ArcGISRuntimeEnvironment.h>
+#include <Error.h>
+#include <ErrorException.h>
+#include <OAuthUserConfiguration.h>
+#include <OAuthUserCredential.h>
+#include <OAuthUserLoginPrompt.h>
+#include <TokenCredential.h>
+
+using namespace Esri::ArcGISRuntime;
+
+namespace Esri::ArcGISRuntime::Toolkit {
+
+/*!
+ \class Toolkit::ArcGISAuthenticationController
+ \inmodule EsriArcGISRuntimeToolkit
+ \ingroup ArcGISQtToolkitUiCppControllers
+ \brief In MVC architecture, this is the controller for the corresponding
+        AuthenticationView.
+ */
+
+/*!
+  \brief Constructor
+  \list
+    \li \a parent Owning QObject.
+  \endlist
+ */
+ArcGISAuthenticationController::ArcGISAuthenticationController(QObject* parent) :
+  ArcGISAuthenticationChallengeHandler(parent)
+{
+  if (!canBeUsed_())
+    return;
+
+  // Set the authentication manager to use this controller
+  ArcGISRuntimeEnvironment::instance()->arcGISAuthenticationManager()->setArcGISAuthenticationChallengeHandler(this);
+
+  // listen for OAuth prompts
+  connect(ArcGISRuntimeEnvironment::instance()->arcGISAuthenticationManager(), &ArcGISAuthenticationManager::oAuthUserLoginPromptIssued, this,
+          [this](OAuthUserLoginPrompt* currentOAuthUserLoginPrompt)
+  {
+    m_currentOAuthUserLoginPrompt = currentOAuthUserLoginPrompt;
+    emit authorizeUrlChanged();
+    emit preferPrivateWebBrowserSessionChanged();
+    emit redirectUrlChanged();
+    emit displayOAuthSignInView();
+  });
+}
+
+/*!
+  \brief Destructor.
+ */
+ArcGISAuthenticationController::~ArcGISAuthenticationController() = default;
+
+ArcGISAuthenticationController* ArcGISAuthenticationController::create(QQmlEngine* qmlEngine, QJSEngine* /*jsEngine*/)
+{
+  static QPointer<ArcGISAuthenticationController> instance = new ArcGISAuthenticationController(qmlEngine);
+
+  if (!instance)
+    instance = new ArcGISAuthenticationController(qmlEngine);
+
+  return instance;
+}
+
+ArcGISAuthenticationController* ArcGISAuthenticationController::instance()
+{
+  return create(nullptr, nullptr);
+}
+
+void ArcGISAuthenticationController::handleArcGISAuthenticationChallenge(ArcGISAuthenticationChallenge* challenge)
+{
+  std::lock_guard<std::mutex> lock(m_mutex);
+  m_currentChallenge = challenge;
+  emit currentAuthenticatingHostChanged();
+
+  // first see if we can handle this with OAuth
+  const auto requestUrl = challenge->requestUrl();
+  for (OAuthUserConfiguration* userConfiguration : std::as_const(m_userConfigurations))
+  {
+    if (userConfiguration->canBeUsedForUrl(requestUrl))
+    {
+      OAuthUserCredential::createAsync(userConfiguration, this).then(this, [this](OAuthUserCredential* credential)
+      {
+        if (!m_currentChallenge)
+          return;
+
+        m_currentChallenge->continueWithCredential(credential);
+        m_currentChallenge->deleteLater();
```

Comment:
Same question about `unique_ptr` here.

Reply:
Here's my concerns. 

`this` is set as context parameter in `then` and `onFailed`. According to the documentation https://doc.qt.io/qt-6/qfuture.html#then-3:

> If the context is destroyed before the chain has finished, the future is canceled.

That means the future can be canceled immediately when the controller is deleted. And in this case, `then` and `onFailed` may not be called and `m_currentChallenge` is not deleted. It looks like it's a memory leak.

To avoid that, I guess it's necessary to do:
- set controller as parent of challenge
- or add `onCancelled` to clean up challenge pointer (and other raw pointers in controller)
- clean up pointers in controller's destructor
- use `unique_ptr`

It's not clear if the controller can be deleted or not. It's implemented as a singleton. But there is a check for the pointer validity in `instance()` so I assume it can be deleted.

Reply:
I see your point and I think it would be better as a std::unique_ptr. I will make that change and retest.